qiskit/algorithms/optimizers/optimizer.py

```diff
@@ -147,6 +149,24 @@ def setting(self):
         ret += f"{params}"
         return ret
 
+    @property
+    @abstractmethod
```

Comment:
Setting this as a required `abstractmethod` feels like a breaking change to me. Unless you don't expect any downstream code to implement custom optimizer classes that are built on this.

Reply:
In this past this was just on the base class and gathered the attributes so you see a summary of the optimizer - it was used by VQE to summarize it configuration. I would not have expected in the past this to be overridden, although you could, because it pulled out all the variables automatically and included them. Now the meaning of this has changed somewhat too now, in that this is a selected sub-set that can be serialized and a sub-class would be expected to override this so things like callable that might be set, which are not serializable, are not included. It still gives a summary in the old sense so its kinda compatible in that regard. Its a choice whether to make this abstract in the base class, or not simply raise a runtime error should it be called on some future optimizer that does not have this. I think the choice was to go with abstract - I had mention this above https://github.com/Qiskit/qiskit-terra/pull/6545#discussion_r661583859

Reply:
But it's true that if a user implemented an `Optimizer` on their own then this would be a breaking change... Maybe it's better to not have it as abstract after all.

Reply:
Having it abstract and an IDE can assist - if its not, and then its not implemented by some sub-class, it will fail at runtime later should the optimizer be serialized. Which is better to have going forwards - abstract or not? If a user has their own optimizer its not a difficult change for them - but if they do then it is a breaking change.

Reply:
Right, my preference is to not make this a required method and remove the `abstractmethod` because of the backwards compatibility issue with any existing downstream optimizer subclasses. The normal pattern for late additions to abstract class like this is to just raise `NotImplementedError` or do a new version with the extra attribute. This doesn't feel to me like something breaking any custom optimizer class over. Especially because the only place this is currently being used is by the ibmq-provider for the qiskit runtime service, so a custom optimizer doesn't need to be serialized for that at all since on the other side the runtime won't be able to instantiate an object from it with out having the custom class in the first place.

But, that being said if people feel super strongly that it needs to be there, I don't think it's necessarily a blocker to have `abstractmethod` set as long it's called out in the release notes as a breaking API change in the `upgrade` section.

Reply:
I made the settings non-abstract, after all the optimizers don't necessarily have to have this implement this, only when they should be sent to the runtime.